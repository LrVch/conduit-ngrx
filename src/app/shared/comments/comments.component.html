<form (ngSubmit)="addComment()">
  <fieldset [disabled]="isSubmitting">
    <div class="card-block">
      <mat-form-field class="input-full-width">
        <textarea matInput
                  rows="3"
                  [formControl]="commentControl"
                  placeholder="Write a comment...">
        </textarea>
        <mat-progress-bar class="comment__add-loader"
                          [hidden]="!isSubmitting"
                          mode="indeterminate">
        </mat-progress-bar>
      </mat-form-field>
    </div>
    <div class="comments__footer">
      <div>
        <img [src]="user.image"
             class="comment-author-img" />
        {{ user.username }}
      </div>
      <button [disabled]="isSubmitting"
              mat-flat-button
              color="primary"
              type="submit">
        Post Comment
      </button>
    </div>

  </fieldset>

  <br>
  <br>
  <br>

  <div [hidden]="!loading"
       class="comments__loader">
    <mat-spinner></mat-spinner>
  </div>

  <app-comment *ngFor="let comment of comments"
               [comment]="comment"
               [canModify]="user.username === comment.author.username"
               [user]="user"
               (delete)="deleteComment(comment)">
  </app-comment>
</form>