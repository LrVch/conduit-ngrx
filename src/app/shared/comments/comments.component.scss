.comments__footer {
    display: flex;
    justify-content: space-between;
    align-items: center;

    img {
        width: 40px;
        height: 40px;
        border-radius: 50%;
    }
}

.comments__loader {
    display: flex;
    justify-content: center;
}

.comment__add-loader {
    position: absolute;
    top: 100%;
    left: 0;
}