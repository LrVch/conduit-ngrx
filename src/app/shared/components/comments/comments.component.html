<app-comment-form *appShowAuthed="true"
                  [user]="user"
                  [reset]="reset"
                  (submitComment)="addComment($event)"
                  [isSubmitting]="isSubmitting"></app-comment-form>

<br>
<br>
<br>

<div [hidden]="!loading"
     class="comments__loader">
  <mat-spinner></mat-spinner>
</div>

<app-comment *ngFor="let comment of comments; trackBy: trackByComment"
             [comment]="comment"
             [locale]="locale"
             [canModify]="user ? user.username === comment.author.username : false"
             (delete)="deleteComment(comment)">