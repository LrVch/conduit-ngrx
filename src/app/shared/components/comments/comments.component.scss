.comments__loader {
  display: flex;
  justify-content: center;
}
