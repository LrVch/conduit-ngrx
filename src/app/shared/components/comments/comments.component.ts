import { Component, Input, Output, EventEmitter } from '@angular/core';
import { User, Comment } from '@app/core';

@Component({
  selector: 'app-comments',
  templateUrl: './comments.component.html',
  styleUrls: ['./comments.component.scss']
})
export class CommentsComponent {
  @Input('locale') locale: string;
  @Input('comments') comments: Comment[] = [];
  @Input('user') user: User = {} as User;
  @Input('loading') loading = false;
  @Input('isSubmitting') isSubmitting = false;
  @Output() delete = new EventEmitter<Comment>();
  @Output() submitComment = new EventEmitter<Comment>();

  @Input('reset') reset = false;

  deleteComment(comment: Comment): void {
    this.delete.emit(comment);
  }

  addComment(comment) {
    this.submitComment.emit(comment);
  }
}
