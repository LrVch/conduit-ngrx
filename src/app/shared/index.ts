export * from './article/article.component';
export * from './articles-list/articles-list.component';
export * from './content-tabs/content-tabs.component';
export * from './tags/tags.component';
export * from './favorite/favorite.component';
export * from './layout';
export * from './shared.module';
export * from './utils';
export * from './list-error/list-error.component';
export * from './show-authed.directive';
export * from './user-info/user-info.component';
export * from './follow-button/follow-button.component';
export * from './tag-list/tag-list.component';
export * from './comments/comments.component';
export * from './base-from/base-from.component';
