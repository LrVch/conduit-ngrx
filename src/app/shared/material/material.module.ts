import { NgModule } from '@angular/core';
import {
  MatButtonModule,
  MatCheckboxModule,
  MatToolbarModule,
  MatCardModule,
  MatChipsModule,
  MatProgressBarModule,
  MatPaginatorModule,
  MatTabsModule,
  MatIconModule,
  MatInputModule,
  MatProgressSpinnerModule,
  MatDialogModule,
  MatSnackBarModule
} from '@angular/material';

const modules = [
  MatButtonModule,
  MatCheckboxModule,
  MatToolbarModule,
  MatCardModule,
  MatChipsModule,
  MatProgressBarModule,
  MatPaginatorModule,
  MatTabsModule,
  MatIconModule,
  MatInputModule,
  MatProgressSpinnerModule,
  MatDialogModule,
  MatSnackBarModule
];

@NgModule({
  imports: modules,
  exports: modules
})
export class MaterialModule { }
