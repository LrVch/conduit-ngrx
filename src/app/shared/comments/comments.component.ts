import { Component, Input, Output, EventEmitter } from '@angular/core';
import { User, Comment } from 'src/app/core';

@Component({
  selector: 'app-comments',
  templateUrl: './comments.component.html',
  styleUrls: ['./comments.component.scss']
})
export class CommentsComponent {
  @Input('comments') comments: Comment[] = [];
  @Input('user') user: User = {} as User;
  @Input('loading') loading = false;
  @Input('isSubmitting') isSubmitting = false;
  @Output() delete = new EventEmitter<Comment>();
  @Output() submitComment = new EventEmitter<Comment>();

  @Input('reset') reset = false;

  deleteComment(comment: Comment): void {
    this.delete.emit(comment);
  }

  addComment(comment) {
    this.submitComment.emit(comment);
  }
}
