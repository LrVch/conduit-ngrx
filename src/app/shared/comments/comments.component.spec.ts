import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests } from '../../lib/testing';
import { CommentsComponent } from './comments.component';
import { DebugElement, Input, EventEmitter, Output, Component } from '@angular/core';
import { By } from '@angular/platform-browser';
import { MaterialModule } from '../material/material.module';
import { User } from 'src/app/core';
import { getUser, getComments } from 'src/app/lib/testing/mock-data.helpers';

@Component({
    selector: 'app-comment-form',
    template: `<div (click)="submit()">a</div>`
})
class CommentFormComponent {
    @Input('user') user: User = {} as User;
    @Input('isSubmitting') isSubmitting = false;
    @Input('reset') reset = false;
    @Output() submitComment = new EventEmitter<string>();

    submit(): void {
        this.submitComment.emit('comment');
    }
}

@Component({
    selector: 'app-comment',
    template: `<div (click)="deleteComment()"></div>`
})
class CommentComponent {
  @Input('comment') comment: Comment = {} as Comment;
  @Input('canModify') canModify = false;
  @Output() delete = new EventEmitter<boolean>();

  deleteComment(): void {
    this.delete.emit(true);
  }
}

describe('CommentsComponent', () => {
    let component: CommentsComponent;
    let fixture: ComponentFixture<CommentsComponent>;
    let de: DebugElement;
    let el: HTMLElement;
    let user;
    let comments;

    beforeEach(
        async(() => {
            user = getUser();
            comments = getComments(3);
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [
                        CommentsComponent,
                        CommentFormComponent,
                        CommentComponent
                    ],
                    imports: [MaterialModule]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(CommentsComponent);
                component = fixture.componentInstance;
                component.comments = comments;
                component.user = user;
                de = fixture.debugElement;
                el = de.nativeElement;
                fixture.detectChanges();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should has "comments" @Input', () => {
        expect(component.comments).toEqual(comments);
    });

    it('should has "user" @Input', () => {
        expect(component.user).toEqual(user);
    });

    it('should has "loading" @Input', () => {
        expect(component.loading).toBe(false);
    });

    it('should has "isSubmitting" @Input', () => {
        expect(component.isSubmitting).toBe(false);
    });

    it('should has "reset" @Input', () => {
        expect(component.reset).toBe(false);
    });

    it('should compile', () => {
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should show "loading" @Input', () => {
        component.loading = true;

        fixture.detectChanges();
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should raise submitComment event when form (submitComment) and pass a coment body', () => {
        let comment = '';
        component.submitComment.subscribe(c => comment = c);

        fixture.detectChanges();

        const submitDe = de.query(By.css('app-comment-form')).query(By.css('div'));

        submitDe.triggerEventHandler('click', null);
        expect(comment).toBe('comment');
    });

    it('should raise deleteComment event when click (delete) on comment and pass a comment', () => {
        let comment = null;
        component.delete.subscribe(c => comment = c);

        fixture.detectChanges();

        const deleteDe = de.query(By.css('app-comment')).query(By.css('div'));
        deleteDe.triggerEventHandler('click', null);

        expect(comment).toEqual(comments[0]);
    });
});

