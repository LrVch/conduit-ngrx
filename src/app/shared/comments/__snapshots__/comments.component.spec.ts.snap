// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsComponent should compile 1`] = `undefined`;

exports[`CommentsComponent should show "loading" @Input 1`] = `
<app-comments
  comments={[Function Array]}
  delete={[Function EventEmitter]}
  isSubmitting="false"
  loading={[Function Boolean]}
  reset="false"
  submitComment={[Function EventEmitter]}
  user={[Function Object]}
>
  <app-comment-form
    ng-reflect-is-submitting="false"
    ng-reflect-reset="false"
    ng-reflect-user="[object Object]"
  >
    <div>
      a
    </div>
  </app-comment-form><br /><br /><br /><div
    class="comments__loader"
  >
    <mat-spinner
      class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation"
      mode="indeterminate"
      role="progressbar"
      style="width: 100px; height: 100px;"
    >
      <svg
        focusable="false"
        ng-reflect-ng-switch="true"
        preserveAspectRatio="xMidYMid meet"
        style="width: 100px; height: 100px;"
        viewBox="0 0 100 100"
      >
        
        <circle
          cx="50%"
          cy="50%"
          r="45"
          style="stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
        />
        
      </svg>
    </mat-spinner>
  </div><app-comment
    ng-reflect-can-modify="false"
    ng-reflect-comment="[object Object]"
  >
    <div />
  </app-comment><app-comment
    ng-reflect-can-modify="false"
    ng-reflect-comment="[object Object]"
  >
    <div />
  </app-comment><app-comment
    ng-reflect-can-modify="false"
    ng-reflect-comment="[object Object]"
  >
    <div />
  </app-comment>
</app-comments>
`;
